pinot-core/src/test/java/org/apache/pinot/core/realtime/impl/fakestream/FakePartitionLevelConsumer.java

```diff
@@ -0,0 +1,121 @@
+import java.util.concurrent.TimeoutException;
+import org.apache.avro.file.DataFileStream;
+import org.apache.avro.generic.GenericDatumWriter;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.BinaryEncoder;
+import org.apache.avro.io.EncoderFactory;
+import org.apache.commons.io.FileUtils;
+import org.apache.pinot.core.realtime.stream.MessageBatch;
+import org.apache.pinot.core.realtime.stream.PartitionLevelConsumer;
+import org.apache.pinot.core.realtime.stream.StreamConfig;
+import org.apache.pinot.core.util.AvroUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.pinot.core.realtime.impl.fakestream.FakeStreamConfigUtils.*;
+
+
+/**
+ * Implementation of {@link PartitionLevelConsumer} for fake stream
+ * Unpacks tar files in /resources/data/On_Time_Performance_2014_partition_<partition>.tar.gz as source of messages
+ */
+public class FakePartitionLevelConsumer implements PartitionLevelConsumer {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(FakePartitionLevelConsumer.class.getName());
+
+  private List<Integer> messageOffsets = new ArrayList<>();
+  private List<byte[]> messageBytes = new ArrayList<>();
+
+  FakePartitionLevelConsumer(int partition, StreamConfig streamConfig) {
+
+    // TODO: this logic can move to a FakeStreamProducer instead of being inside the Consumer
+    File tempDir = new File(FileUtils.getTempDirectory(), getClass().getSimpleName());
+    File outputDir = new File(tempDir, String.valueOf(partition));
+
+    int offset = 0;
+
+    try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream(65536)) {
+      File avroFile = unpackAvroTarFile(outputDir).get(0);
+
+      int numPartitions = FakeStreamConfigUtils.getNumPartitions(streamConfig);
+
+      try (DataFileStream<GenericRecord> reader = AvroUtils.getAvroReader(avroFile)) {
+        BinaryEncoder binaryEncoder = new EncoderFactory().directBinaryEncoder(outputStream, null);
+        GenericDatumWriter<GenericRecord> datumWriter = new GenericDatumWriter<>(reader.getSchema());
+
+        int recordNumber = 0;
+        for (GenericRecord genericRecord : reader) {
+          if (getPartitionNumber(recordNumber++, numPartitions) != partition) {
+            continue;
+          }
+          outputStream.reset();
+
+          datumWriter.write(genericRecord, binaryEncoder);
+          binaryEncoder.flush();
+
+          byte[] bytes = outputStream.toByteArray();
+          // contiguous offsets
+          messageOffsets.add(offset++);
+          messageBytes.add(bytes);
+        }
+      }
+    } catch (Exception e) {
+      LOGGER.error("Could not create {}", FakePartitionLevelConsumer.class.getName(), e);
+    } finally {
+      FileUtils.deleteQuietly(outputDir);
+    }
+  }
+
+  @Override
+  public MessageBatch fetchMessages(long startOffset, long endOffset, int timeoutMillis) throws TimeoutException {
+    if (startOffset >= FakeStreamConfigUtils.getLargestOffset()) {
```

Comment:
Instead of unpacking to a tempdir, is it possible to just include the loop in this place, and get the messages from the original avro file? If this does not seem straight-forward, you can add that as a TODO and work on it later. Just trying to make tests go faster, and try not to fill out tempdir.

Reply:
Thanks for clarifying, @npawar . We cannot do the proposed method since we want to return records based on an offset.